Adds "Disable Stripe" confirmation modal

brettshumaker opened the pull request:
# Changes proposed in this Pull Request:

Fixes #1681

This PR adds a "disable Stripe" confirmation modal when a user attempts to disable the Stripe payment method when UPE is enabled and they have 1 or more payment methods added or express checkouts are enabled. If the user cancels the modal, the gateway remains enabled, if they confirm it the gateway is disabled.

# Testing instructions

From `/wp-admin/admin.php?page=wc-settings&tab=checkout`:
1. With UPE **disabled**, click the toggle to disable the `Stripe` method - no modal should appear.

From `/wp-admin/admin.php?page=wc-settings&tab=checkout`:
1. With UPE **enabled**, click `Stripe UPE` to manage settings.
2. Remove all payment methods.
3. Disable `Express Checkouts`.
4. Go back to `/wp-admin/admin.php?page=wc-settings&tab=checkout`
5. Click the toggle to disable the `Stripe` method - no modal should appear.

From `/wp-admin/admin.php?page=wc-settings&tab=checkout`:
1. With UPE **enabled**, click `Stripe UPE` to manage settings.
2. Remove all payment methods.
3. Enable `Express Checkouts`.
4. Go back to `/wp-admin/admin.php?page=wc-settings&tab=checkout`
5. Click the toggle to disable the `Stripe` method - the confirmation modal should appear.

From `/wp-admin/admin.php?page=wc-settings&tab=checkout`:
1. With UPE **enabled**, click `Stripe UPE` to manage settings.
2. Add at least 1 payment method.
4. Go back to `/wp-admin/admin.php?page=wc-settings&tab=checkout`
5. Click the toggle to disable the `Stripe` method - the confirmation modal should appear.

brettshumaker committed:
Initial pass at adding the "disable Stripe" confirmation modal.

I was able to use a lot of what was done on WCPay for a similar feature. I'll be going back through and checking for WCPay references in what I've brought over, updating @since tags, etc.

brettshumaker committed:
Moved the conditional to load the modal code to the main plugin file, updated @since tags

brettshumaker committed:
Also handle the case where no payment methods exist, but express payments are enabled.

brettshumaker committed:
Add test that the `DisableConfirmationModal` calls the correct handlers.

frosso commented on client/payment-gateways/__tests__/index.test.js:
```diff
@@ -0,0 +1,38 @@
+/**
+ * External dependencies
+ */
```
We don't need the `External dependencies`/`Internal dependencies` in the imports, they can all be removed from these files
```suggestion
```

brettshumaker replied:
Noted. Removed them everywhere I had them.

frosso commented on client/payment-gateways/style.scss:
```diff
@@ -0,0 +1,34 @@
+	}
+}
+
+.wcstripe-confirmation-modal {
+
+	&.alert {
```
Just so that the spacing between blocks is applied consistently

```suggestion
.wcstripe-confirmation-modal {
	&.alert {
```

brettshumaker replied:
Ah, yep. Missed that - thank you.

frosso commented on woocommerce-gateway-stripe.php:
```diff
@@ -207,6 +208,13 @@ public function init() {
 					}
+
+					// If UPE is enabled and there are enabled payment methods, we need to load the disable Stripe confirmation modal.
+					$enabled_upe_payment_methods = get_option( 'woocommerce_stripe_settings', [] )['upe_checkout_experience_accepted_payments'];
+					$upe_payment_requests_enabled = 'yes' === get_option( 'woocommerce_stripe_settings', [] )['payment_request'];
```
Moving these checks (with the exclusion of `WC_Stripe_Feature_Flags::is_upe_checkout_enabled()`, which seems to be applied in these files) to the `includes/admin/class-wc-stripe-admin.php` file would help colocate the logic, what do you think?

It could be placed in the constructor (before registering the hooks) or on each hook as well 🤷

brettshumaker replied:
Yep...that makes sense. I had them in the other file originally, but then moved them here before making the PR. It makes sense, though, to split them out to checking if UPE is enabled in this file, THEN check in the included file if the other conditions are met to need the modal.

frosso commented on includes/admin/class-wc-stripe-admin.php:
```diff
@@ -0,0 +1,71 @@
+			'WC_STRIPE_PAYMENT_GATEWAYS_PAGE',
+			$payment_gateways_script_src_url,
+			$payment_gateways_script_asset['dependencies'],
+			WC_Stripe::get_file_version( 'build/payment_gateways.js' ),
```
The `*.asset.php` files inside the `build` directory already contain a `version` attribute - should we use that instead?
Example: https://github.com/woocommerce/woocommerce-gateway-stripe/blob/baa79e5c199a3dd0145437cea849ecca645be1ab/includes/admin/class-wc-stripe-settings-controller.php#L48-L67

brettshumaker replied:
Hah! Absolutely. This is what happens when I'm borrowing the bulk of code from another plugin and fail to look around in the codebase I'm bringing it to.

frosso commented on includes/admin/class-wc-stripe-admin.php:
```diff
@@ -0,0 +1,71 @@
+
+}
+
+new WC_Stripe_Admin();
```
This is creating the side effect of registering the hooks the moment the file is included.
Adding a `new` here also makes it hard to possibly pass any arguments to the constructor for dependency injection when instantiating the class.

What do you think about instantiating the class after the file's inclusion, in `woocommerce-gateway-stripe.php` instead?

brettshumaker replied:
Yes, I'll do that instead and follow the existing pattern.

frosso commented on webpack.config.js:
```diff
@@ -40,5 +40,6 @@ module.exports = {
 		upe_onboarding_wizard: './client/upe-onboarding-wizard/index.js',
 		upe_opt_in_banner: './client/entrypoints/upe-opt-in-banner/index.js',
 		upe_settings: './client/settings/index.js',
+		payment_gateways: './client/payment-gateways/index.js',
```
We started placing entrypoints in the `entrypoints` directory - see pc4etw-jM-p2

brettshumaker replied:
👍 Moved my files to `./client/entrypoints/payment-gateways`

frosso commented on includes/admin/class-wc-stripe-admin.php:
```diff
@@ -0,0 +1,71 @@
+ *
+ * @since 5.6.0
+ */
+class WC_Stripe_Admin {
```
The name `WC_Stripe_Admin` seems a bit generic for something whose purpose if of enqueuing the script for the payments tab.
What do you think about renaming this to something less generic? Maybe `WC_Stripe_Payment_Gateways_Controller` (so that it matches the name of the asset that it enqueues) or `WC_Stripe_Payments_Page_Controller` (or something similar - I'm open to other possibilities)?

brettshumaker replied:
Agreed - I like `WC_Stripe_Payment_Gateways_Controller`.

frosso reviewed (commented):
Looks functionally sound! A few suggestions on the consistency of the code, otherwise looks good!

brettshumaker committed:
Remove internal/external dependency blocks

brettshumaker committed:
Make scss spacing more consistent

brettshumaker committed:
Addressing more review feedback.

- updated and moved part of the logic for whether or not to include the files needed for the modal.
- use the version attribute already provided by the `*.asset.php` file in the build directory
- instantiate the new controller immediately after including the file instead of directly in the file itself.
- Renamed the class to be more specific to its function.
- Moved the new component files into the `client/entrypoints` folder.

brettshumaker committed:
Updated path in test

brettshumaker committed:
Merge branch 'develop' into add/disable-stripe-confirmation-modal

frosso commented on includes/admin/class-wc-stripe-payment-gateways-controller.php:
```diff
@@ -0,0 +1,80 @@
+	 */
+	public function __construct() {
+		// If UPE is enabled and there are enabled payment methods, we need to load the disable Stripe confirmation modal.
+		$enabled_upe_payment_methods = get_option( 'woocommerce_stripe_settings', [] )['upe_checkout_experience_accepted_payments'];
+		$upe_payment_requests_enabled = 'yes' === get_option( 'woocommerce_stripe_settings', [] )['payment_request'];
```
(minor)
```suggestion
        $stripe_settings = get_option( 'woocommerce_stripe_settings', [] );
		$enabled_upe_payment_methods = $stripe_settings['upe_checkout_experience_accepted_payments'];
		$upe_payment_requests_enabled = 'yes' === $stripe_settings['payment_request'];
```

But it's probably going to need some linting :D

brettshumaker replied:
Fixed!

frosso commented on woocommerce-gateway-stripe.php:
```diff
@@ -602,6 +608,21 @@ public function register_routes() {
 					$stripe_account_keys_controller->register_routes();
 				}
 			}
+
+			/**
+			 * Gets the file modified time as a cache buster if we're in dev mode, or the plugin version otherwise.
+			 *
+			 * @since 5.6.0
+			 * @param string $file Local path to the file.
+			 * @return string The cache buster value to use for the given file.
+			 */
+			public static function get_file_version( $file ) {
+				if ( defined( 'SCRIPT_DEBUG' ) && SCRIPT_DEBUG ) {
+					$file = trim( $file, '/' );
+					return filemtime( WC_STRIPE_ABSPATH . $file );
+				}
+				return WC_STRIPE_VERSION;
+			}
```
It looks like this isn't used anywhere - remove?
```suggestion
```

brettshumaker replied:
D'oh! Removed.

frosso reviewed (approved):
Thanks for the changes, looks good!

brettshumaker committed:
Fixing up those other minor points.

brettshumaker committed:
Merge branch 'develop' into add/disable-stripe-confirmation-modal
